### ISSUE TITLE
local_vars_dump with recursive data structures causes infinite recursion

### ISSUE THREAD

**Issue by jlesueur** [rollbar/rollbar-php#410]
Unfortunately some of our parameters are recursive. Here is a super simplified example:
```php
$obj = [
  'key1' => 'value',
  'key2' => 'value'
];

$obj['key3'] => &$obj;
```

This causes `Utility::serialize($obj)` to enter an infinite recursion. So if local_vars_dump is on, errors can cause infinite recursion.

I kind of think a simple fix is to implement a maximum depth for the serialize method. Maybe a configurable one, with a default of somewhere between 5 and 10? If that solution sounds ok, I'd be willing to put together a PR for it.

**Comment by jessewgibbs** [comment]
@jlesueur thanks for reporting this issue.

@ArturMoczulski what do you think about the proposal to set a maximum depth for the `serialize()` method?  I imagine we'd want to have it be a configurable setting...

**Comment by coryvirok** [comment]
We also need the serialization code to detect loops. See the other sdks for a reference implementation.

**Comment by ArturMoczulski** [comment]
@jessewgibbs yeah I think that's good

@jlesueur if you want to prepare a PR that would be awesome

@coryvirok I'll check out other SDKs, although I believe introducing default max depth should solve the problem

**Comment by coryvirok** [comment]
It might solve it in terms of the specific bug but the serialization will
be incorrect due to loops. Also, it will have another effect for nested
objects of increasing the payload size.

**Comment by JohnKendhammer** [comment]
Issue has been resolved with PR #414
